docs/notebooks/code_overview.pct.py

```diff
@@ -0,0 +1,238 @@
+#
+# Multiple models can also be combined into a single `ModelStack` model that combines their outputs for prediction and sampling. This can be useful when modelling multi-objective observations with independent models. There are also constructors like `TrainableModelStack` and `PredictJointModelStack` that combine specific types of model for use with code that requires that type, delegating the relevant methods where appropriate.
+
+# %% [markdown]
+# ### `SearchSpace`
+#
+# The `SearchSpace` base class represents the domain over which the objective function is to be optimized. Spaces can currently be either continuous `Box` spaces, discrete `DiscreteSearchSpace` spaces, or a `TaggedProductSearchSpace` product of multipe spaces. All search spaces expose their dimensions, bounds and a sampler.
+
+# %% [markdown]
+# ### `AcquisitionRule`
+#
+# The `AcquisitionRule` base class represents a routine for selecting new query points during a Bayesian optimization loop (via an `acquire` method). It is generic on three types:
+#
+# * **ResultType**: the output of the rule, typically this is just tensors representing the query points. However, it can also be functions that accept some *acquisition state* and return the query points with a new state.
+# * **SearchSpaceType**: the type of the search space; any optimizer that the rule uses must support this.
+# * **ProbabilisticModelType**: the type of the models; any acquisition functions or samplers that the rule uses must support this.
+#
+# Exampes of rules include:
+#
+# 1. `EfficientGlobalOptimization` (EGO) is the most commonly used rule, and uses acquisition functions and optimizers to select new query points.
+# 1. `AsynchronousOptimization` is similar to EGO but uses acquisition state to keep track of points whose observations are still pending.
+# 1. `DiscreteThompsonSampling` uses Thompson samplers rather than acquisition functions to select new query points.
+#
+
+# %% [markdown]
+# ### `AcquisitionFunction` and `AcquisitionFunctionBuilder`
+#
+# The `AcquisitionFunction` type definition represents any acquisition function: that is, a function that maps a set of query points to a single value that describes how useful it would be evaluate all these points together.
+#
+# The `AcquisitionFunctionBuilder` base class, meanwhile, represents something that builds and updates acquisition functions. At the start of the Bayesian optimization, the builder's `prepare_acquisition_function` method is called by the acquisition rule to create an acquisition function from the current observations and probabilistic models. For efficiency, most builders also define an `update_acquisition_function` method for updating the function using the updated observations and models. (The ones that don't instead generate a new acquisition function when necessary.)
+#
+# Acquisition functions that support only one probabilistic model are more easily defined using the `SingleModelAcquisitionBuilder` convenience class.
```

Comment:
could be good to give a bit more detail about this? why it is convenient, what it changes...